test/openssl/test_provider.rb

```diff
@@ -12,14 +12,33 @@ def test_openssl_provider_name_inspect
   end
 
   def test_openssl_provider_names
+    # We expect the following providers are loaded in the cases:
+    # * Non-FIPS: default
+    # * FIPS: fips, base
+    # Use the null provider to test the added provider.
+    # See provider(7) - OPENSSL PROVIDERS to see the list of providers, and
+    # OSSL_PROVIDER-null(7) to check the details of the null provider.
     with_openssl <<-'end;'
-      base_provider = OpenSSL::Provider.load("base")
-      assert_equal(2, OpenSSL::Provider.provider_names.size)
-      assert_includes(OpenSSL::Provider.provider_names, "base")
+      num = OpenSSL::Provider.provider_names.size
 
-      assert_equal(true, base_provider.unload)
-      assert_equal(1, OpenSSL::Provider.provider_names.size)
-      assert_not_includes(OpenSSL::Provider.provider_names, "base")
+      added_provider = OpenSSL::Provider.load("null")
+      assert_equal(num + 1, OpenSSL::Provider.provider_names.size)
+      assert_includes(OpenSSL::Provider.provider_names, "null")
+
+      assert_equal(true, added_provider.unload)
+      assert_equal(num, OpenSSL::Provider.provider_names.size)
+      assert_not_includes(OpenSSL::Provider.provider_names, "null")
+    end;
+  end
+
+  def test_openssl_provider_names_on_fips
+    omit "Only for FIPS" unless OpenSSL.fips_mode
+
+    # The number of providers is at least 2 in FIPS. They are the fips provider
+    # for the key management, and another provider for the encoding/decoding.
+    with_openssl(<<-'end;')
+      assert_compare(2, "<=", OpenSSL::Provider.provider_names.size)
+      assert_include(OpenSSL::Provider.provider_names, "fips")
```

Comment:
This seems to be true in OpenSSL 3.0-3.3, but is it part of the API guarantee that the FIPS mode loads two providers?

To be honest I don't get what failure modes `test_openssl_provider_names_on_fips` intends to cover, as `test_openssl_provider_names` already works in the FIPS mode.

Reply:
> This seems to be true in OpenSSL 3.0-3.3, but is it part of the API guarantee that the FIPS mode loads two providers?

Yes. To run OpenSSL in FIPS properly, we need to load the fips provider and also another provider base or default at least. I was told this by someone at OpenSSL in the past in the OpenSSL project. Below is the document, though it doesn't explain it clearly.

https://docs.openssl.org/master/man7/fips_module/#making-all-applications-use-the-fips-module-by-default

> Note that this configuration also activates the "base" provider. The base provider does not include any cryptographic algorithms (and therefore does not impact the validation status of any cryptographic operations), but does include other supporting algorithms that may be required. It is designed to be used in conjunction with the FIPS module.

> To be honest I don't get what failure modes test_openssl_provider_names_on_fips intends to cover, as test_openssl_provider_names already works in the FIPS mode.

The method `OpenSSL.fips_mode` that we rely on to judge if FIPS is enable or not is the `EVP_default_properties_is_fips_enabled`.

https://github.com/ruby/openssl/blob/ade50769dfac81c994a5c7fe64884faf7d75ae44/ext/openssl/ossl.c#L412

According to the EVP_default_properties_is_fips_enabled(3), the C function also checks the property `'fips=yes'`, and it doesn't check if fips and (base or/and default) providers are loaded to run with OpenSSL FIPS properly.

https://docs.openssl.org/master/man3/EVP_set_default_properties/

> EVP_default_properties_is_fips_enabled() indicates if 'fips=yes' is a default property for the given libctx.

So, my intention for this test `test_openssl_provider_names_on_fips` is to check the proper providers to run with OpenSSL FIPS properly.

Does it make sense to you?

Reply:
I updated the comment lines adding the referred man pages.

Reply:
> https://docs.openssl.org/master/man7/fips_module/#making-all-applications-use-the-fips-module-by-default

Thanks for the pointer. I wasn't sure if the base provider is part of the public API or an implementation detail (which could change in any version, breaking our tests) since we don't usually use the base provider directly. Users are indeed required to explicitly load either the base or default provider to use OpenSSL in the FIPS mode.

I understood these assertions are correct, but:

> So, my intention for this test `test_openssl_provider_names_on_fips` is to check the proper providers to run with OpenSSL FIPS properly.

I think checking the user's OpenSSL configuration is out of scope for `test_provider.rb` when it doesn't increase test coverage of OpenSSL::Provider. Besides, I suspect in a situation where these assertions fail, the test case can't run at all because `require "openssl"` wouldn't work.

I thought https://github.com/ruby/openssl/pull/783 was intended for this exact purpose.

Reply:
> I think checking the user's OpenSSL configuration is out of scope for `test_provider.rb` when it doesn't increase test coverage of OpenSSL::Provider. Besides, I suspect in a situation where these assertions fail, the test case can't run at all because `require "openssl"` wouldn't work.
> 
> I thought #783 was intended for this exact purpose.

All right. Your points make sense to me. We don't see the failure case actually as you pointed out, because the `require "openssl"` wouldn't work in the case. And it doesn't also increase the test coverage. So, I removed the `test_openssl_provider_names_on_fips`.